File: training/java/src/main/java/de/sots/cellarsandcentaurs/adapter/persistence/Repository.java
package de.sots.cellarsandcentaurs.adapter.persistence;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

public abstract class Repository<T> {
    private final Map<UUID, T> rows = new HashMap<>();

    protected abstract UUID idOf(T row);

    public void save(T row) {
        rows.put(idOf(row), row);
    }

    public Optional<T> findById(UUID id) {
        return Optional.ofNullable(rows.get(id));
    }

    public Collection<T> findAll() {
        return rows.values();
    }
}

File: training/php/src/Adapter/Persistence/Repository.php
<?php

declare(strict_types=1);

namespace De\Sots\CellarsAndCentaurs\Adapter\Persistence;

use Countable;

/**
 * @template T of object
 */
abstract class Repository implements Countable
{
    /** @var array<string, T> */
    private array $items = [];

    /** @param T $item */
    public function save(object $item): void
    {
        $this->items[$this->idOf($item)] = $item;
    }

    /** @return T|null */
    public function findOne(string $id): ?object
    {
        return $this->items[$id] ?? null;
    }

    /** @return list<T> */
    public function findAll(): array
    {
        return array_values($this->items);
    }

    public function count(): int
    {
        return count($this->items);
    }

    /** @param T $item */
    abstract protected function idOf(object $item): string;
}
